# Pandas: grouping

In [1]:
import pandas as pd
import numpy as np

In [2]:
cars = pd.read_csv("vehicles.csv")

In [3]:
cars.head()

Make                Model  Year  Engine Displacement  \
0        AM General    DJ Po Vehicle 2WD  1984                  2.5   
1        AM General     FJ8c Post Office  1984                  4.2   
2        AM General  Post Office DJ5 2WD  1985                  2.5   
3        AM General  Post Office DJ8 2WD  1985                  4.2   
4  ASC Incorporated                  GNX  1987                  3.8   

   Cylinders     Transmission        Drivetrain                Vehicle Class  \
0        4.0  Automatic 3-spd     2-Wheel Drive  Special Purpose Vehicle 2WD   
1        6.0  Automatic 3-spd     2-Wheel Drive  Special Purpose Vehicle 2WD   
2        4.0  Automatic 3-spd  Rear-Wheel Drive  Special Purpose Vehicle 2WD   
3        6.0  Automatic 3-spd  Rear-Wheel Drive  Special Purpose Vehicle 2WD   
4        6.0  Automatic 4-spd  Rear-Wheel Drive                 Midsize Cars   

  Fuel Type  Fuel Barrels/Year  City MPG  Highway MPG  Combined MPG  \
0   Regular          19.388824        18           17            17   
1   Regular          25.354615        13           13            13   
2   Regular          20.600625        16           17            16   
3   Regular          25.354615        13           13            13   
4   Premium          20.600625        14           21            16   

   CO2 Emission Grams/Mile  Fuel Cost/Year  
0               522.764706            1950  
1               683.615385            2550  
2               555.437500            2100  
3               683.615385            2550  
4               555.437500            2550

How many Car models?

In [4]:
cars["Model"].nunique()

3608

We have 3608 unique car models.

group by the data by the Make using count function

In [5]:
cars_grouped_by_make = cars.groupby(["Make"]).count()
cars_grouped_by_make

Model  Year  Engine Displacement  Cylinders  \
Make                                                                       
AM General                       4     4                    4          4   
ASC Incorporated                 1     1                    1          1   
Acura                          302   302                  302        302   
Alfa Romeo                      41    41                   41         41   
American Motors Corporation     22    22                   22         22   
...                            ...   ...                  ...        ...   
Volkswagen                    1047  1047                 1047       1047   
Volvo                          717   717                  717        717   
Wallace Environmental           32    32                   32         32   
Yugo                             8     8                    8          8   
smart                           20    20                   20         20   

                             Transmission  Drivetrain  Vehicle Class  \
Make                                                                   
AM General                              4           4              4   
ASC Incorporated                        1           1              1   
Acura                                 302         302            302   
Alfa Romeo                             41          41             41   
American Motors Corporation            22          22             22   
...                                   ...         ...            ...   
Volkswagen                           1047        1047           1047   
Volvo                                 717         717            717   
Wallace Environmental                  32          32             32   
Yugo                                    8           8              8   
smart                                  20          20             20   

                             Fuel Type  Fuel Barrels/Year  City MPG  \
Make                                                                  
AM General                           4                  4         4   
ASC Incorporated                     1                  1         1   
Acura                              302                302       302   
Alfa Romeo                          41                 41        41   
American Motors Corporation         22                 22        22   
...                                ...                ...       ...   
Volkswagen                        1047               1047      1047   
Volvo                              717                717       717   
Wallace Environmental               32                 32        32   
Yugo                                 8                  8         8   
smart                               20                 20        20   

                             Highway MPG  Combined MPG  \
Make                                                     
AM General                             4             4   
ASC Incorporated                       1             1   
Acura                                302           302   
Alfa Romeo                            41            41   
American Motors Corporation           22            22   
...                                  ...           ...   
Volkswagen                          1047          1047   
Volvo                                717           717   
Wallace Environmental                 32            32   
Yugo                                   8             8   
smart                                 20            20   

                             CO2 Emission Grams/Mile  Fuel Cost/Year  
Make                                                                  
AM General                                         4               4  
ASC Incorporated                                   1               1  
Acura                                            302             302  
Alfa Romeo                                        41              41 

Converting Grams/Mile to Grams/Km

1 Mile = 1.60934 Km

Converting Gallons to Liters

1 Gallon = 3.78541 Liters

What brand has the most cars?

In [6]:
cars_grouped_by_make["Model"].sort_values(ascending=False).head(1)

Make
Chevrolet    3643
Name: Model, dtype: int64

In [7]:
# option 2
cars["Make"].value_counts().idxmax()

'Chevrolet'

1. **show the average CO2_Emission_Grams/Km by Brand**
2. **show the average CO2_Emission_Grams/Km by Brand ... sorted**

In [8]:
avg_by_brand = cars.groupby(["Make"]) \
["CO2 Emission Grams/Mile"].mean().reset_index(name="Avg CO2 Emission (Grams/Mile)") \
.sort_values(by="Avg CO2 Emission (Grams/Mile)",ascending=False).reset_index(drop=True)

In [9]:
avg_by_brand["Avg CO2 Emission (Grams/Km)"]=round(avg_by_brand["Avg CO2 Emission (Grams/Mile)"]*0.621371,2)
avg_by_brand.drop(columns="Avg CO2 Emission (Grams/Mile)", inplace=True)

avg_by_brand

Make  Avg CO2 Emission (Grams/Km)
0                                Vector                       651.92
1      Superior Coaches Div E.p. Dutton                       552.21
2    S and S Coach Company  E.p. Dutton                       552.21
3                               Bugatti                       542.50
4                Laforza Automobile Inc                       502.01
..                                  ...                          ...
122                                MINI                       194.93
123                            Daihatsu                       192.74
124                                Fiat                       189.31
125                               smart                       153.50
126                              Fisker                       105.01

[127 rows x 2 columns]


- Use pd.cut or pd.qcut to create 4 groups (bins) of cars, by Year. We want to explore how cars have evolved decade by decade.

In [10]:
year_bins = [1980,1990,2000,2010,2020]
cars["Year_Range"] = pd.cut(cars["Year"],bins=year_bins)

## Did cars consume more gas in the eighties?

show the average City_Km/Liter by year_range

In [11]:
avg = cars.groupby(["Year_Range"])[["City MPG"]].mean()
avg["average City_Km/Liter"] = round(avg["City MPG"]*(1.60934/3.78541),2)
avg.drop(columns="City MPG", inplace=True)

In [12]:
avg

average City_Km/Liter
Year_Range                         
(1980, 1990]                   7.35
(1990, 2000]                   7.20
(2000, 2010]                   7.21
(2010, 2020]                   8.39

## We can see that cars in the 90s consumed the most.

Which brands are more environment friendly?

In [13]:
env_friendly = cars.groupby(["Year_Range","Make"]) \
["CO2 Emission Grams/Mile"].mean().reset_index(name="CO2 Emission Grams/Mile")

In [14]:
env_friendly["CO2 Emission Grams/Km"] = round(env_friendly["CO2 Emission Grams/Mile"]/1.60934,2)
env_friendly.drop(columns="CO2 Emission Grams/Mile",inplace=True)
env_friendly.sort_values(by="CO2 Emission Grams/Km")

Year_Range                         Make  CO2 Emission Grams/Km
417  (2010, 2020]                       Fisker                 105.01
380  (2000, 2010]                        smart                 153.39
507  (2010, 2020]                        smart                 153.54
26   (1980, 1990]                     Daihatsu                 182.82
416  (2010, 2020]                         Fiat                 189.31
..            ...                          ...                    ...
500  (2010, 2020]                       Vector                    NaN
501  (2010, 2020]          Vixen Motor Company                    NaN
502  (2010, 2020]  Volga Associated Automobile                    NaN
505  (2010, 2020]        Wallace Environmental                    NaN
506  (2010, 2020]                         Yugo                    NaN

[508 rows x 3 columns]

### Fisker, Smart, Daihatsu and Fiat are the most environment friendly.

Does the drivetrain affect fuel consumption?

In [15]:
drive_train = cars.groupby(["Drivetrain"])[["Highway MPG","City MPG"]].mean()
drive_train["Highway_Km/Liter"] = drive_train["Highway MPG"]*(1.60934/3.78541)
drive_train["City_Km/Liter"] = drive_train["City MPG"]*(1.60934/3.78541)
drive_train.drop(columns=["Highway MPG","City MPG"],inplace=True)
drive_train

Highway_Km/Liter  City_Km/Liter
Drivetrain                                                 
2-Wheel Drive                       8.222444       6.642480
2-Wheel Drive, Front               14.029714      10.628571
4-Wheel Drive                       9.668584       7.190861
4-Wheel or All-Wheel Drive          8.347130       6.392049
All-Wheel Drive                    10.882531       7.785598
Front-Wheel Drive                  12.166210       9.002214
Part-time 4-Wheel Drive             8.115385       6.215696
Rear-Wheel Drive                    9.023946       6.556574

## The front consumes fuels the most. 

Do cars with automatic transmission consume more fuel than cars with manual transmission?

In [16]:
cars["Transmission"].unique()

array(['Automatic 3-spd', 'Automatic 4-spd', 'Manual 5-spd',
       'Automatic (S5)', 'Manual 6-spd', 'Automatic 5-spd', 'Auto(AM8)',
       'Auto(AM-S8)', 'Auto(AV-S7)', 'Automatic (S6)', 'Automatic (S9)',
       'Automatic (S4)', 'Auto(AM-S9)', 'Automatic (S7)', 'Auto(AM7)',
       'Auto(AM-S7)', 'Auto(AM6)', 'Automatic 6-spd', 'Manual 4-spd',
       'Automatic (S8)', 'Manual(M7)', 'Auto(AM-S6)',
       'Automatic (variable gear ratios)', 'Automatic (AV)',
       'Auto(AV-S8)', 'Automatic (AM6)', 'Automatic 8-spd', 'Auto(A1)',
       'Automatic (A1)', 'Automatic (A6)', 'Auto(AV-S6)', 'Manual 3-spd',
       'Manual 7-spd', 'Automatic 9-spd', 'Auto (AV)', 'Automatic 6spd',
       'Auto(L4)', 'Auto(L3)', 'Auto (AV-S6)', 'Auto (AV-S8)',
       'Automatic (AV-S6)', 'Automatic 7-spd', 'Manual 5 spd',
       'Auto(AM5)', 'Automatic (AM5)'], dtype=object)

In [17]:
def unify_transmission(transmission):
    if "Auto" in transmission:
        return 'Automatic'
    elif 'Manual' in transmission:
        return 'Manuel'
    else:
        return np.nan

cars['Transmission'] = cars['Transmission'].apply(unify_transmission)

In [19]:
cars["Transmission"].unique()

array(['Automatic', 'Manuel'], dtype=object)

In [37]:
uni = cars.groupby("Transmission").agg({"City MPG":"mean"}).apply(lambda x: x*(1.60934/3.78541)). \
rename(columns={"City MPG":"City_Km/Liter"}).sort_values(by="City_Km/Liter",ascending=False)
uni

City_Km/Liter
Transmission               
Manuel             7.968348
Automatic          7.278292

## **Cars with Manuel transmission consume more.**

Use groupby and agg with different aggregation measures for different columns:

aggregate with average City_Km/Liter and the count of the Trans

In [67]:
unni = cars.groupby(["Transmission"]).agg({"City MPG":"mean","Transmission":"count"}).apply(lambda x: x*(1.60934/3.78541)). \
rename(columns={"City MPG":"City_Km/Liter"})
unni["Transmission"]= unni["Transmission"]/((1.60934/3.78541))
unni["Transmission"]= unni["Transmission"].astype("int64")
unni

City_Km/Liter  Transmission
Transmission                             
Automatic          7.278292         24290
Manuel             7.968348         11661